test/unit/c_api/eigensolver/test_gen_eigensolver_c_api.cpp

```diff
@@ -0,0 +1,261 @@
+//
+// Distributed Linear Algebra with Future (DLAF)
+//
+// Copyright (c) 2018-2023, ETH Zurich
+// All rights reserved.
+//
+// Please, refer to the LICENSE file in the root directory.
+// SPDX-License-Identifier: BSD-3-Clause
+//
+
+#include <functional>
+#include <tuple>
+
+#include <pika/init.hpp>
+
+#include <dlaf/communication/communicator_grid.h>
+#include <dlaf/eigensolver/gen_eigensolver.h>
+#include <dlaf/matrix/copy.h>
+#include <dlaf/matrix/matrix.h>
+#include <dlaf_c_test/c_api_helpers.h>
+
+#include "test_gen_eigensolver_c_api_wrapper.h"
+
+#include <gtest/gtest.h>
+
+#include <dlaf_test/comm_grids/grids_6_ranks.h>
+#include <dlaf_test/eigensolver/test_gen_eigensolver_correctness.h>
+#include <dlaf_test/matrix/util_matrix.h>
+
+using namespace dlaf;
+using namespace dlaf::comm;
+using namespace dlaf::matrix;
+using namespace dlaf::test;
+using namespace testing;
+
+::testing::Environment* const comm_grids_env =
+    ::testing::AddGlobalTestEnvironment(new CommunicatorGrid6RanksCAPIEnvironment);
+
+template <typename Type>
+class GenEigensolverTest : public TestWithCommGrids {};
+
+template <class T>
+using GenEigensolverTestMC = GenEigensolverTest<T>;
+
+TYPED_TEST_SUITE(GenEigensolverTestMC, MatrixElementTypes);
+
+#ifdef DLAF_WITH_GPU
+template <class T>
+using GenEigensolverTestGPU = GenEigensolverTest<T>;
+
+TYPED_TEST_SUITE(GenEigensolverTestGPU, MatrixElementTypes);
+#endif
+
+const std::vector<blas::Uplo> blas_uplos({blas::Uplo::Lower});
+
+const std::vector<std::tuple<SizeType, SizeType, SizeType>> sizes = {
+    // {m, mb, eigensolver_min_band}
+    {0, 2, 100},                                              // m = 0
+    {5, 8, 100}, {34, 34, 100},                               // m <= mb
+    {4, 3, 100}, {16, 10, 100}, {34, 13, 100}, {32, 5, 100},  // m > mb
+    {34, 8, 3},  {32, 6, 3}                                   // m > mb, sub-band
+};
+
+template <class T, Backend B, Device D, API api>
+void testGenEigensolver(const blas::Uplo uplo, const SizeType m, const SizeType mb,
+                        CommunicatorGrid grid, int dlaf_context) {
+  // In normal use the runtime is resumed by the C API call
+  // The pika runtime is suspended by dlaf_initialize
+  // Here we need to resume it manually to build the matrices with DLA-Future
+  pika::resume();
+
+  const TileElementSize block_size(mb, mb);
+
+  auto create_reference = [&]() -> auto{
```

Comment:
Very minor, is the `auto` needed?
```suggestion
  auto create_reference = [&]() {
```

Reply:
I don't think so. It was copy-pasted from `test_gen_eigensolver.cpp`, so I removed it there as well.